File: web/src/services/collection.service.js
import db from "../db/database";

export const collectionService = {
  async create(data) {
    const id = crypto.randomUUID();
    const now = new Date().toISOString();

    const collection = {
      id,
      user_id: data.user_id,
      name: data.name.trim(),
      description: data.description?.trim() || null,
      is_default: data.is_default || false,
      is_ai_generated: data.is_ai_generated || false,
      created_at: now,
      updated_at: now,
      is_deleted: false,
    };

    await db.transaction("rw", [db.collections, db.sync_queue], async () => {
      await db.collections.put(collection);
      await db.sync_queue.add({
        table_name: "collections",
        record_id: id,
        operation: "CREATE",
        payload: collection,
        created_at: now,
        synced: false,
      });
    });

    return collection;
  },

  async getAll(userId) {
    const collections = await db.collections
      .where("user_id")
      .equals(userId)
      .and((c) => !c.is_deleted)
      .toArray();

    // Attach word count
    for (const col of collections) {
      const links = await db.vocabulary_collections
        .where("collection_id")
        .equals(col.id)
        .and((vc) => !vc.is_deleted)
        .toArray();
      col.word_count = links.length;
    }

    // Sort: default first, then alphabetically
    collections.sort((a, b) => {
      if (a.is_default && !b.is_default) return -1;
      if (!a.is_default && b.is_default) return 1;
      return a.name.localeCompare(b.name);
    });

    return { items: collections };
  },

  async getById(id) {
    const collection = await db.collections.get(id);
    if (!collection || collection.is_deleted) return null;

    const vcLinks = await db.vocabulary_collections
      .where("collection_id")
      .equals(id)
      .and((vc) => !vc.is_deleted)
      .toArray();
    const vocabIds = vcLinks.map((vc) => vc.vocabulary_id);

    let vocabularies = [];
    if (vocabIds.length > 0) {
      vocabularies = await db.vocabularies
        .where("id")
        .anyOf(vocabIds)
        .and((v) => !v.is_deleted)
        .toArray();

      for (const vocab of vocabularies) {
        vocab.definitions = await db.definitions
          .where("vocabulary_id")
          .equals(vocab.id)
          .and((d) => !d.is_deleted)
          .sortBy("sort_order");
      }
    }

    return { collection, vocabularies };
  },

  async update(id, updates) {
    const now = new Date().toISOString();
    const updated = { ...updates, updated_at: now };
    delete updated.id;

    await db.transaction("rw", [db.collections, db.sync_queue], async () => {
      await db.collections.update(id, updated);
      await db.sync_queue.add({
        table_name: "collections",
        record_id: id,
        operation: "UPDATE",
        payload: updated,
        created_at: now,
        synced: false,
      });
    });

    return db.collections.get(id);
  },

  async delete(id) {
    const now = new Date().toISOString();

    await db.transaction(
      "rw",
      [db.collections, db.vocabulary_collections, db.sync_queue],
      async () => {
        await db.collections.update(id, { is_deleted: true, updated_at: now });

        const vcLinks = await db.vocabulary_collections
          .where("collection_id")
          .equals(id)
          .toArray();
        for (const vc of vcLinks) {
          await db.vocabulary_collections.update(vc.id, {
            is_deleted: true,
            updated_at: now,
          });
        }

        await db.sync_queue.add({
          table_name: "collections",
          record_id: id,
          operation: "DELETE",
          payload: { is_deleted: true },
          created_at: now,
          synced: false,
        });
      },
    );
  },

  async assignWord(vocabularyId, collectionId) {
    const id = crypto.randomUUID();
    const now = new Date().toISOString();

    const link = {
      id,
      vocabulary_id: vocabularyId,
      collection_id: collectionId,
      created_at: now,
      updated_at: now,
      is_deleted: false,
    };

    await db.transaction(
      "rw",
      [db.vocabulary_collections, db.sync_queue],
      async () => {
        await db.vocabulary_collections.put(link);
        await db.sync_queue.add({
          table_name: "vocabulary_collections",
          record_id: id,
          operation: "CREATE",
          payload: link,
          created_at: now,
          synced: false,
        });
      },
    );
  },

  async removeWord(vocabularyId, collectionId) {
    const now = new Date().toISOString();

    const link = await db.vocabulary_collections
      .where("vocabulary_id")
      .equals(vocabularyId)
      .and((vc) => vc.collection_id === collectionId && !vc.is_deleted)
      .first();

    if (!link) return;

    await db.transaction(
      "rw",
      [db.vocabulary_collections, db.sync_queue],
      async () => {
        await db.vocabulary_collections.update(link.id, {
          is_deleted: true,
          updated_at: now,
        });
        await db.sync_queue.add({
          table_name: "vocabulary_collections",
          record_id: link.id,
          operation: "DELETE",
          payload: { is_deleted: true },
          created_at: now,
          synced: false,
        });
      },
    );
  },
};

File: mobile/lib/services/collection_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';

import '../db/database.dart' as db;
import '../models/collection.dart';

class CollectionService {
  final SupabaseClient? _client;
  final db.AppDatabase _db;

  CollectionService({
    SupabaseClient? client,
    required db.AppDatabase database,
  })  : _client = client,
        _db = database;

  SupabaseClient get _supabase => _client ?? Supabase.instance.client;

  /// Reactive stream of user collections with active word counts
  Stream<List<db.CollectionWithCount>> watchAll(String userId) {
    return _db.collectionDao.watchAll(userId);
  }

  /// Snapshot of user collections with active word counts
  Future<List<db.CollectionWithCount>> getAll(String userId) async {
    return await _db.collectionDao.getAll(userId);
  }

  /// Get single collection by ID from local cache
  Future<db.Collection?> getById(String id) async {
    return await _db.collectionDao.getById(id);
  }

  /// Create collection: writes to Supabase, then updates Drift cache
  Future<Collection> create({
    required String userId,
    required String name,
    String? description,
    bool isDefault = false,
    bool isAiGenerated = false,
  }) async {
    final row = await _supabase
        .from('collections')
        .insert({
          'user_id': userId,
          'name': name.trim(),
          'description': description?.trim(),
          'is_default': isDefault,
          'is_ai_generated': isAiGenerated,
        })
        .select()
        .single();

    final created = Collection.fromJson(row);
    await _db.collectionDao.upsert(created.toDriftCompanion());
    return created;
  }

  /// Update collection: writes to Supabase, then updates Drift cache
  Future<void> update(Collection collection) async {
    final now = DateTime.now();
    await _supabase.from('collections').update({
      'name': collection.name.trim(),
      'description': collection.description?.trim(),
      'updated_at': now.toIso8601String(),
    }).eq('id', collection.id);

    await _db.collectionDao.upsert(collection.toDriftCompanion());
  }

  /// Soft-delete collection and unbind words on Supabase and Drift cache
  Future<void> delete(String id) async {
    final now = DateTime.now().toIso8601String();

    await _supabase.from('collections').update({
      'is_deleted': true,
      'updated_at': now,
    }).eq('id', id);

    await _supabase.from('vocabulary_collections').update({
      'is_deleted': true,
      'updated_at': now,
    }).eq('collection_id', id);

    await _db.collectionDao.deleteCollection(id);
  }

  /// Assign word to collection
  Future<void> assignWord(String vocabularyId, String collectionId) async {
    await _supabase.from('vocabulary_collections').upsert(
      {
        'vocabulary_id': vocabularyId,
        'collection_id': collectionId,
        'is_deleted': false,
        'updated_at': DateTime.now().toIso8601String(),
      },
      onConflict: 'vocabulary_id,collection_id',
    );

    await _db.collectionDao.assignToCollection(vocabularyId, collectionId);
  }

  /// Remove word from collection
  Future<void> removeWord(String vocabularyId, String collectionId) async {
    await _supabase.from('vocabulary_collections').update({
      'is_deleted': true,
      'updated_at': DateTime.now().toIso8601String(),
    }).match({
      'vocabulary_id': vocabularyId,
      'collection_id': collectionId,
    });

    await _db.collectionDao.removeFromCollection(vocabularyId, collectionId);
  }

  /// Get collection IDs for a specific vocabulary from local cache
  Future<List<String>> getCollectionIdsForVocabulary(
    String vocabularyId,
  ) async {
    return await _db.collectionDao.getCollectionIdsForVocabulary(vocabularyId);
  }
}
